content/academy/puppeteer_playwright/reading_intercepting_requests.md

```diff
@@ -0,0 +1,352 @@
+
+    console.log('Request for followers was made!')
+});
+</marked-tab>
+```
+
+> Note that you should always define any request reading/interception code prior to calling the `page.goto()` function.
+
+Cool! So now when we run our code, we'll see this logged to the console:
+
+```text
+Request for followers was made!
+```
+
+This request includes some useful query parameters, namely the `client_id`. Let's go ahead and grab these values from the request URL and print them to the console:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+// Listen for all requests
+page.on('request', (req) => {
+    // If the URL doesn't include our keyword, ignore it
+    if (!req.url().includes('followings')) return;
+
+    // Convert the request URL into a URL object
+    const url = new URL(req.url());
+
+    // Print the search parameters in object form
+    console.log(Object.fromEntries(url.searchParams));
+});
+
+await page.goto('https://soundcloud.com/tiesto/following');
+
+await page.waitForTimeout(10000);
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+// Listen for all requests
+page.on('request', (req) => {
+    // If the URL doesn't include our keyword, ignore it
+    if (!req.url().includes('followings')) return;
+
+    // Convert the request URL into a URL object
+    const url = new URL(req.url());
+
+    // Print the search parameters in object form
+    console.log(Object.fromEntries(url.searchParams));
+});
+
+await page.goto('https://soundcloud.com/tiesto/following');
+
+await page.waitForTimeout(10000);
+await browser.close();
+</marked-tab>
+```
+
+After running this code, we can see this logged to the console:
+
+```text
+{
+  client_id: 'llCGDUjKpxUslgO1yEce7Zh95PXE78Bo',
+  limit: '12',
+  offset: '0',
+  linked_partitioning: '1',
+  app_version: '1652347025',
+  app_locale: 'en'
+}
+```
+
+## [](#reading-responses) Reading responses
+
+Listening for and reading responses is very similar to reading requests. The only difference is that we need to listen for the **response** event instead of **request**. Additionally, the object passed into the callback function represents the response instead of the request.
+
+This time, instead of grabbing the query parameters of the request URL, let's grab hold of the response body and print it to the console in JSON format:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+// Notice that the callback function is now async
+page.on('response', async (res) => {
+    if (!res.request().url().includes('followings')) return;
+
+    // Grab the response body in JSON format
+    const json = await res.json();
+    console.log(json);
+});
```

Comment:
The res.json() can throw if the response is not JSON, causing an unhandled promise rejection. Having unhandled errors or rejections in the page handlers is quite a common error of devs, so I think it would make sense to add a try catch and use that opportunity to explain why that's important.